Bump into the latest glTF loader

takahirox opened the pull request:
Requires: https://github.com/MozillaReality/three.js/pull/57
From: #4117

This PR replaces our Three.js glTF loader fork with the latest official Three.js one and updates `gltf-model-plus.js` and `avatar-editor.js` to follow the latest official Three.js `GLTFLoader` API.

Benefits

* Less cost for maintenance because we will no longer need to maintain our GLTFLoader fork
* We will be easily able to add or manage new and existing glTF related features with the official GLTFLoader plugin API for example LOD

Please refer to https://github.com/mozilla/hubs/issues/4117#issuecomment-818319656 for the changes.

Note that the changes in this PR is just for replacing the glTF loader. There may be a chance that they can be further cleaned up or optimized with the new API. But I would like to do them if needed in another PR because I don't want to make this PR bigger and the review harder.

TODO
- [x] merge https://github.com/MozillaReality/three.js/pull/57
- [x] update package.json to point back at hubs/master for three
- [x] update package-lock.json to point back at hubs/master and the merged commit sha

takahirox commented on package.json:
```diff
@@ -134,7 +134,7 @@
     "screenfull": "^4.0.1",
     "semver": "^7.3.2",
-    "three": "github:mozillareality/three.js#hubs/master",
+    "three": "github:mozillareality/three.js#hubs/r128GLTFLoader",
```
If https://github.com/MozillaReality/three.js/pull/57 is merged, I would update this line back to `master` in this PR or follow up PR. `package-lock.json` is same.

netpro2k replied:
I usually do it just before merging, and I leave a [ ] todo in the PR description to remind me

takahirox commented:
What I have tested so far.
- Enter a room and see an environment is rendered (it means the loader succeeds in loading a glTF scene)
- Upload a glTF object including moz light map extension
- See avatar preview in Avatar settings
- Upload a custom avatar and change the avatar to it

netpro2k commented on src/components/gltf-model-plus.js:
```diff
@@ -341,127 +340,221 @@ function runMigration(version, json) {
+  afterRoot(gltf) {
+    gltf.scene.traverse(object => {
+      // GLTFLoader sets matrixAutoUpdate on animated objects, we want to keep the defaults
+      // @TODO: Should this be fixed in the gltf loader?
```
I know this was just copied over from our existing code, but it still does seem like the GLTFLoader should be respecting this default.

takahirox replied:
I think `GLTFLoader` should respect the default configuration, too. I'm going to suggest it to Three.js at some point. If it will be merged, we can drop this code.

netpro2k commented on src/components/gltf-model-plus.js:
```diff
@@ -341,127 +340,221 @@ function runMigration(version, json) {
+    // When we replace our Three.js fork with the latest official one
+    // we need to revisit this workaround.
+    if (!parser.textureLoader.isImageBitmapLoader && typeof createImageBitmap !== undefined) {
+      parser.textureLoader = new THREE.ImageBitmapLoader(parser.options.manager);
```
ImageBitmapLoader returns an ImageBitmap not a Texture instance I thought. Doesn't this need to be a some kind of texture loader?

takahirox replied:
[`GLTFLoader` makes `CanvasTexture` from `ImageBitmap` if `textureLoader` is `ImageBitmapLoader`](https://github.com/mrdoob/three.js/blob/r128/examples/jsm/loaders/GLTFLoader.js#L2578-L2586) so this line should be fine.

But this line may be confusing for people who don't know the inside of `GLTFLoader`. I would add comment.

netpro2k commented on src/components/gltf-model-plus.js:
```diff
@@ -341,127 +340,221 @@ function runMigration(version, json) {
+        return;
+      }
+      const materials = Array.isArray(object.material) ? object.material : [object.material];
+      // @TODO: Handle more efficiently
```
We are already overriding `parser.textureLoader` in the constructor. Seems like we can still use the HubsTextureLoader to take care of this no?

takahirox replied:
I didn't really want to override `parser.textureLoader` because `GLTFLoader` doesn't have a proper API to override the texture loader. Overriding `parser.textureLoader = new CustomLoader()` is hacky and the `GLTFLoader` doesn't expect the texture loader to be overridden. There might be a chance that it causes future errors especially when updating to the future glTF loader.

I did override the texture loader because there doesn't seem to be other options to let `GLTFLoader` use `ImageBitmapLoader` for Firefox. Yes, overriding the texture loader is hacky as I wrote but `GLTFLoader` expects that the texture loader can be `ImageBitmapLoader` so I think it would be less problematic than using our custom texture loader. And if Firefox and Three.js will be updated to support `ImageBitmapLoader` we will be able to simply drop this hacky code.

But if we realize that this traverse is slow I want to revisit.

netpro2k commented on src/components/gltf-model-plus.js:
```diff
@@ -341,127 +340,221 @@ function runMigration(version, json) {
-      node.extras.gltfIndex = i;
+    //
+    if (this.fileMap) {
+      // The GLTF is now cached as a THREE object, we can get rid of the original blobs
```
Just a mental note I re-remember while reading through this PR. This is to clean up the root level resources we create (from Sketchfab zip files) to load the GLTF.

Does the official GLTFLoader have any method of cleaning up the resources it creates?

takahirox replied:
If a plugin manage resources `afterRoot` hook would be a good place to release the resources.

netpro2k commented on src/components/gltf-model-plus.js:
```diff
@@ -341,127 +340,221 @@ function runMigration(version, json) {
-  }
+    //
+    let version = 0;
+    if (
```
Not particularly important but eventually we might want this to be its own MozHubsComponetnsExtension that just handles the components extension and not the other weird hubs specific workarounds. Doesn't really matter too much now since the components stuff is all still very tightly integrated into aframe and the rest of Hubs.

takahirox replied:
Added a comment so far and would like to think of the cleaning up later.

netpro2k commented on src/components/gltf-model-plus.js:
```diff
@@ -341,127 +340,221 @@ function runMigration(version, json) {
 
-      if (!node.extras) {
-        node.extras = {};
+    // Replace animation target node name with the node uuid.
```
As we have discussed before, it still seems a bit weird that this isn't just the default behavior since GLTF does not guarantee unique names... This is definitely preferable to maintaining a fork, but would like to get to the bottom of why this isn't always desirable.

netpro2k commented on src/components/gltf-model-plus.js:
```diff
@@ -341,127 +340,221 @@ function runMigration(version, json) {
+  //        Cons is loadMaterial hook is fired as _invokeOne so
+  //        if other plugins defining loadMaterial is registered
+  //        there is a chance that this light map extension handler isn't called.
+  //        The glTF loader should be updated to remove the limitation.
```
Hmm yeah we should discuss. Figuring out ow multiple extensions handling the same thing should interact is tricky, but I could definitely see cases where there are multiple layers of extensions for loading a single thing.

netpro2k commented on src/components/gltf-model-plus.js:
```diff
@@ -341,127 +340,221 @@ function runMigration(version, json) {
+  //        So using loadMaterial hook as workaround so far.
+  //        Cons is loadMaterial hook is fired as _invokeOne so
+  //        if other plugins defining loadMaterial is registered
+  //        there is a chance that this light map extension handler isn't called.
```
Since we have control over the loader this shouldnt really be an issue in practice, but we will need to be careful when doing upgrades that we don't accidentally break it by some plugin being introduced upstream. Hopefully this workaround doesn't last too long anyway.

netpro2k commented on src/components/gltf-model-plus.js:
```diff
@@ -341,127 +340,221 @@ function runMigration(version, json) {
+    }
+
+    if (!parser.options.ktx2Loader) {
+      // @TODO: Display warning (only if the extension is in extensionsRequired)?
```
In practice we will always be providing a ktx2loader so if this happens its certainly a bug... We probably don't want to fail silently.

takahirox replied:
OK, I would add warning here

netpro2k commented on src/components/gltf-model-plus.js:
```diff
@@ -341,127 +340,221 @@ function runMigration(version, json) {
 
-      node.extras.gltfIndex = i;
+    //
+    if (this.fileMap) {
```
Might be cleaner to just handle this in `loadGLTF` instead of passing `fileMap` through.

takahirox replied:
Sounds good. Updated.

netpro2k commented on src/components/gltf-model-plus.js:
```diff
@@ -569,7 +662,7 @@ AFRAME.registerComponent("gltf-model-plus", {
       this.disposeLastInflatedEl();
 
-      this.model = gltf.scene || gltf.scenes[0];
+      this.model = gltf.scene;
```
Is this guaranteed to be set now? I think the reason we had it originally is that not GLTFs are not required to have a default scene (though almost all I have encountered in practice do).

takahirox replied:
[Yes, it's guaranteed in the latest `GLTFLoader`](https://github.com/mrdoob/three.js/blob/r128/examples/jsm/loaders/GLTFLoader.js#L2042). IIRC the official loader didn't do that before so I think you encountered the problem with the old loader.

netpro2k commented on src/react-components/avatar-editor.js:
```diff
@@ -47,6 +47,53 @@ const fetchAvatar = async avatarId => {
+    const { body } = parser.extensions.KHR_binary_glTF;
+    const content = JSON.stringify(ensureAvatarMaterial(parser.json));
+
+    // Inject hubs components on upload. Used to create base avatar
```
I think you can just get rid of this commented code now. This predates the Blender extension and was an easy way for us to create some of the initial avatars, but we don't need it anymore.

takahirox replied:
Done, thanks

netpro2k commented on src/react-components/avatar-editor.js:
```diff
@@ -47,6 +47,53 @@ const fetchAvatar = async avatarId => {
+    // This plugin just wants to split gltf and bin from glb and
+    // doesn't want to start the parse. But glTF loader plugin API
+    // doesn't have an ability to cancel the parse. So overriding
+    // parser.json with very light glTF data as workaround.
```
hah nice workaround

takahirox replied:
Good to know that this workaround seems acceptable.

netpro2k commented on src/react-components/avatar-editor.js:
```diff
@@ -113,42 +160,18 @@ class AvatarEditor extends Component {
-      this.inputFiles.bin = new File([body], "file.bin", {
-        type: "application/octet-stream"
+      await new Promise((resolve, reject) => {
+        // This GLTFBinarySplitterPlugin saves gltf and bin in inputFiles.gltf/bin
```
Something about this feels a bit weird but I don't really have a much better idea. I maybe would have had the plugin just store it on itself rather than have it passed in and then done something like this after load:

```js
this.inputFiles.gltf = binarySplitterPlugin.gltf;
this.inputFiles.bin = binarySplitterPlugin.bin;
```

But not sure that is really much better.

takahirox replied:
How about this?

```
await new Promise((resolve, reject) => {
  // GLTFBinarySplitterPlugin saves gltf and bin in gltf.files
  gltfLoader.load(gltfUrl, result => {
    this.inputFiles.gltf = result.files.gltf;
    this.inputFiles.bin = result.files.bin;
  }, onProgress, reject);
});
```

netpro2k reviewed (commented):
This looks great! Definitely feels like a lot more of this is now being don the "right way'. Still some hacks in place but I think we have a clear path forward to improve most of them. Nice work!

brianpeiris commented:
I will remove my need to review, since you are both already experts in this code and Dom has done a thorough review.

takahirox commented:
Thanks for the reviews.

Just in case, I would like to inform you folks that please don't merge this PR until we finish the test with the QA team.

takahirox commented:
I just noticed (remembered) that the latest `GLTFLoader` has dropped DDS texture support because `MSFT_texture_dds` extension is a vendor extension, not a Khronos KHR extension, and they encourage to use more standard compressed texture `KHR_texture_basisu` extension now.

If we want to keep supporting DDS texture, we can use an external DDS texture plugin I wrote 

https://github.com/takahirox/three-gltf-extensions/tree/main/loaders/MSFT_texture_dds

Any opinion on this? We should encourage KHR_texture_basisu, too, but also I think we don't need to drop DDS texture support and require the asset authors to update their works.

takahirox commented:
Ah wait. I noticed that the current `gltf-model-plus.js` doesn't set `DDSLoader` to `GLTFLoader`. That means we don't support DDS texture even now. Then no DDS texture support won't be problem.

takahirox commented:
I reverted the removal of `HubsTextureLoader` from `gltf-model-plus.js` because it removes the resolved texture image url from Three.js Cache and I couldn't come up with a workaround with the more proper API like plugin API.

As we internally discussed on Discord, we may disable Three.js entire cache so we may want to revisit later.

I added a comment about why we use `HubsTextureLoader` and we should replace it later so far.

takahirox commented:
Somehow the glTF loader replacement resolves the Specular/Glossiness material problem #1613 The Specular/Glossiness material is rendered correctly on the testing server.

takahirox commented:
Regarding [`parser.cache.removeAll()`](https://github.com/mozilla/hubs/blob/ebfe192502f32cfda00fce22b8b675b65891cb4e/src/components/gltf-model-plus.js#L448), I thought I added it in `GLTFHubsPlugin` but it seems I somehow accidentally delete it.

I think it's ok because `glTFLoader` and `parser` aren't saved anywhere so the resources used by them should be garbage collected automatically.

https://github.com/mozilla/hubs/blob/693dc621377734a5391e78e18c4d7a2c947b169d/src/components/gltf-model-plus.js#L515-L539
https://github.com/mozilla/hubs/blob/693dc621377734a5391e78e18c4d7a2c947b169d/src/components/gltf-model-plus.js#L638-L718
